File: db/src/main/java/onl/ycode/stormify/NamingPolicy.java
// SPDX-License-Identifier: Apache-2.0
// (C) Panayotis Katsaloulis

package onl.ycode.stormify;

import static onl.ycode.stormify.Utils.camelToSnake;

/**
 * Naming policy for converting class names to table names and field names to column names.
 * It is possible to create custom policies or use the predefined ones.
 */
public interface NamingPolicy {
    /**
     * Convert a class or field name to a table or column name.
     * By using this method, it is possible to define rules in a central location
     * and free the objects of Annotating manually.
     *
     * @param s The given Java name
     * @return The converted Database name
     */
    String convert(String s);

    /**
     * Use the name as is.
     */
    NamingPolicy camelCase = s -> s;
    /**
     * Convert the name to lower case with underscores (snake_case).
     */
    NamingPolicy lowerCaseWithUnderscores = s -> camelToSnake(s, false);
    /**
     * Convert the name to upper case with underscores (SCREAMING_SNAKE_CASE).
     */
    NamingPolicy upperCaseWithUnderscores = s -> camelToSnake(s, true);
}

File: stormify/src/commonMain/kotlin/onl/ycode/stormify/NamingPolicy.kt
// SPDX-License-Identifier: Apache-2.0
// (C) Panayotis Katsaloulis
package onl.ycode.stormify

/**
 * Built-in naming policies for converting Kotlin property/class names to database column/table names.
 *
 * A naming policy is simply a `(String) -> String` function. Use one of the presets below
 * (e.g. `NamingPolicy.LOWER_CASE_WITH_UNDERSCORES`) or assign your own lambda directly:
 *
 * ```kotlin
 * stormify.namingPolicy = { "tbl_${it.lowercase()}" }
 * ```
 */
object NamingPolicy {
    /** Keeps names as-is (camelCase). */
    val CAMEL_CASE: (String) -> String = { it }

    /** Converts camelCase to lower_case_with_underscores (snake_case). This is the default. */
    val LOWER_CASE_WITH_UNDERSCORES: (String) -> String = { camelToSnake(it, false) }

    /** Converts camelCase to UPPER_CASE_WITH_UNDERSCORES (SCREAMING_SNAKE_CASE). */
    val UPPER_CASE_WITH_UNDERSCORES: (String) -> String = { camelToSnake(it, true) }
}

internal fun camelToSnake(str: String, upper: Boolean): String {
    val result = StringBuilder()
    for (c in str) {
        if (c.isUpperCase() && result.isNotEmpty()) {
            result.append('_')
            result.append(if (upper) c.uppercaseChar() else c.lowercaseChar())
        } else
            result.append(if (upper) c.uppercaseChar() else c.lowercaseChar())
    }
    return result.toString()
}
